### ISSUE TITLE
RLVR arguments clarification

### ISSUE THREAD

**Issue by hank0316** [allenai/open-instruct#524]
Dear Tulu Team,

Recently, I want to explore RLVR on various RL algorithms, such as RLOO and REINFORCE++, using the [OpenRLHF repository](https://github.com/OpenRLHF/OpenRLHF) implementation. However, I am having difficulty understanding the argument name mapping between Open-Instruct and OpenRLHF.

Specifically, I would like clarification on the following points related to Table 20:

1. Does the "Mini-batches $N_{mb}$" in Table 20 indicate that the mini-batch size is set to 1 on each GPU? I was confused since the argument `local_mini_batch_size` is set to 32 in the [reproduction command](https://github.com/allenai/open-instruct/blob/main/docs/tulu3.md#llama-31-tulu-3-8b-reproduction).  
2. In OpenRLHF, the argument `num_episodes` specifies the number of iterations over the entire dataset (e.g., setting num_episodes to 10 results in iterating through the dataset 10 times). Does the "Total Episodes" in Table 20 have the same meaning?
3. In the [reproduction command](https://github.com/allenai/open-instruct/blob/main/docs/tulu3.md#llama-31-tulu-3-8b-reproduction), the `total_episodes` is set to 10000000, which differs from the value in Table 20 of the report. Is this a typo, or does the argument have a different meaning compared to Table 20?
4. In the [reproduction command](https://github.com/allenai/open-instruct/blob/main/docs/tulu3.md#llama-31-tulu-3-8b-reproduction), the `per_device_train_batch_size` is set to 2, while Table 20 states that the effective batch size is 512. Does this imply that 512 / 2 = 256 GPUs were used for PPO training on the 8B model?

Thank you for your time and assistance. I look forward to your response.

Best regards,
Tzu-Han

**Comment by vwxyzjn** [comment]
Hi @hank0316, thanks for raising the question. 

1. The `Mini-batches` actually refers to `num_mini_batches`. Here is how the for loop goes. Though thanks to your issue, I realized we didn't actually update the `num_mini_batches` properly according to `local_mini_batch_size`, so I will submit a fix later.

```
import numpy as np
batch_size = 8
nminibatches = 2
gradient_accumulation_steps = 2
mini_batch_size = batch_size // nminibatches
micro_batch_size = mini_batch_size // gradient_accumulation_steps
data = np.arange(batch_size).astype(np.float32)
print("data:", data)
print("batch_size:", batch_size)
print("mini_batch_size:", mini_batch_size)
print("micro_batch_size:", micro_batch_size)
for epoch in range(4):
    batch_inds = np.random.permutation(batch_size)
    print("epoch:", epoch, "batch_inds:", batch_inds)
    for mini_batch_start in range(0, batch_size, mini_batch_size):
        mini_batch_end = mini_batch_start + mini_batch_size
        mini_batch_inds = batch_inds[mini_batch_start:mini_batch_end]
        
        # `optimizer.zero_grad()` set optimizer to zero for gradient accumulation
        for micro_batch_start in range(0, mini_batch_size, micro_batch_size):
            micro_batch_end = micro_batch_start + micro_batch_size 
            micro_batch_inds = mini_batch_inds[micro_batch_start:micro_batch_end]
            print("____⏩ a forward pass on", data[micro_batch_inds])
        # `optimizer.step()`
        print("⏪ a backward pass on", data[mini_batch_inds])

# data: [0. 1. 2. 3. 4. 5. 6. 7.]
# batch_size: 8
# mini_batch_size: 4
# micro_batch_size: 2
# epoch: 0 batch_inds: [6 4 0 7 3 5 1 2]
# ____⏩ a forward pass on [6. 4.]
# ____⏩ a forward pass on [0. 7.]
# ⏪ a backward pass on [6. 4. 0. 7.]
# ____⏩ a forward pass on [3. 5.]
# ____⏩ a forward pass on [1. 2.]
# ⏪ a backward pass on [3. 5. 1. 2.]
# epoch: 1 batch_inds: [6 7 3 2 0 4 5 1]
# ____⏩ a forward pass on [6. 7.]
# ____⏩ a forward pass on [3. 2.]
# ⏪ a backward pass on [6. 7. 3. 2.]
# ____⏩ a forward pass on [0. 4.]
# ____⏩ a forward pass on [5. 1.]
# ⏪ a backward pass on [0. 4. 5. 1.]
# epoch: 2 batch_inds: [1 4 5 6 0 7 3 2]
# ____⏩ a forward pass on [1. 4.]
# ____⏩ a forward pass on [5. 6.]
# ⏪ a backward pass on [1. 4. 5. 6.]
# ____⏩ a forward pass on [0. 7.]
# ____⏩ a forward pass on [3. 2.]
# ⏪ a backward pass on [0. 7. 3. 2.]
# epoch: 3 batch_inds: [7 2 4 1 3 0 6 5]
# ____⏩ a forward pass on [7. 2.]
# ____⏩ a forward pass on [4. 1.]
# ⏪ a backward pass on [7. 2. 4. 1.]
# ____⏩ a forward pass on [3. 0.]
# ____⏩ a forward pass on [6. 5.]
# ⏪ a backward pass on [3. 0. 6. 5.]
```
2. No. `total_episodes` refers to how many prompt-completion pairs we used throughout the entire training. For example, say your dataset has 10 prompt-completion pairs, and we set `total_episodes=40`, then we are effectively training for 40 epochs. Using episodes as the x-axis has a rich history. E.g., 

* https://arxiv.org/pdf/1909.08593, 
<img width="710" alt="Image" src="https://github.com/user-attachments/assets/3150ac7c-9744-497a-b5e3-225c6f5f1336" />

* https://arxiv.org/abs/2009.01325
<img width="706" alt="Image" src="https://github.com/user-attachments/assets/52fbd018-f377-48d2-b142-b2c7fbb337f8" />

3. It's not a typo. We basically copied and pasted the command as much as we could. Note that we did not take the final checkpoint but an intermediate checkpoint as the released model, so we effectively trained for ~50k `total_episodes`.

<img width="1319" alt="Image" src="https://github.com/user-attachments/assets/3bb1f816-97ed-4422-8cdf-94baf58cfc57" />

<img width="1370" alt="Image" src="https://github.com/user-attachments/assets/015bb671-6619-4f8d-a2c4-1e10bba317a9" />

4. No, we did not use 256 GPUs for this. We just used 8 GPUs and we can simulate the larger batch size with gradient accumulation. Note that `local_mini_batch_size = 32`. Since we use 7 GPUs for training, the global batch size would be 7 * 32 = 224. Table 20 has the incorrect number and we will update it in the next arxiv update.

**Comment by hank0316** [comment]
Got it. Thank you for your detailed reply!
